crates/orchestrator/src/discovery/monitor.rs

```diff
@@ -87,215 +118,75 @@ impl DiscoveryMonitor {
-        let nodes = nodes
-            .into_iter()
-            .filter(|node| node.is_validated)
-            .collect::<Vec<DiscoveryNode>>();
-
-        Ok(nodes)
-    }
-
-    pub async fn fetch_nodes_from_discovery(&self) -> Result<Vec<DiscoveryNode>, Error> {
-        let mut all_nodes = Vec::new();
-        let mut any_success = false;
-
-        for discovery_url in &self.discovery_urls {
-            match self.fetch_nodes_from_single_discovery(discovery_url).await {
-                Ok(nodes) => {
-                    info!(
-                        "Successfully fetched {} nodes from {}",
-                        nodes.len(),
-                        discovery_url
-                    );
-                    all_nodes.extend(nodes);
-                    any_success = true;
-                }
-                Err(e) => {
-                    error!("Failed to fetch nodes from {discovery_url}: {e}");
-                }
-            }
-        }
-
-        if !any_success {
-            error!("Failed to fetch nodes from all discovery services");
-            return Ok(Vec::new());
-        }
-
-        // Remove duplicates based on node ID
-        let mut unique_nodes = Vec::new();
-        let mut seen_ids = std::collections::HashSet::new();
-        for node in all_nodes {
-            if seen_ids.insert(node.node.id.clone()) {
-                unique_nodes.push(node);
-            }
-        }
-
-        info!(
-            "Total unique nodes after deduplication: {}",
-            unique_nodes.len()
-        );
-        Ok(unique_nodes)
-    }
-
-    async fn count_healthy_nodes_with_same_endpoint(
+    async fn count_healthy_nodes_with_same_peer_id(
         &self,
         node_address: Address,
-        ip_address: &str,
-        port: u16,
-    ) -> Result<u32, Error> {
+        peer_id: &p2p::PeerId,
+    ) -> Result<u32> {
         let nodes = self.store_context.node_store.get_nodes().await?;
         Ok(nodes
             .iter()
             .filter(|other_node| {
                 other_node.address != node_address
-                    && other_node.ip_address == ip_address
-                    && other_node.port == port
+                    && other_node.p2p_id == peer_id.to_string()
                     && other_node.status == NodeStatus::Healthy
             })
             .count() as u32)
     }
 
-    async fn sync_single_node_with_discovery(
-        &self,
-        discovery_node: &DiscoveryNode,
-    ) -> Result<(), Error> {
-        let node_address = discovery_node.node.id.parse::<Address>()?;
+    async fn perform_node_updates(&self, node: &NodeWithMetadata) -> Result<()> {
+        let node_address = node.node().id.parse::<Address>()?;
 
-        // Check if there's any healthy node with the same IP and port
-        let count_healthy_nodes_with_same_endpoint = self
-            .count_healthy_nodes_with_same_endpoint(
+        // Check if there's any healthy node with the same peer ID
```

Comment:
this was changed from ip:port collisions to p2p peer ID collisions under the assumption that distinct nodes will have distinct peer IDs, if two nodes have the same peer ID they're actually the same node.